### PR TITLE
Use big transactions for small recordsize writes.

### PR DESCRIPTION
**Pull request by amotin**
When ZFS appends files in chunks bigger than recordsize, it borrows buffer from ARC and fills it before opening transaction.  This supposed to help in case of page faults to not hold transaction open indefinitely.  The problem appears when recordsize is set lower than default 128KB. Since each block is committed in separate transaction, per-transaction overhead becomes significant, and what is even worse, active use of of per-dataset and per-pool locks to protect space use accounting for each transaction badly hurts the code SMP scalability.  The same transaction size limitation applies in case of file rewrite, but without even excuse of buffer borrowing.

To address the issue, disable the borrowing mechanism if recordsize is smaller than default and the write request is 4+x bigger than it. In such case writes up to 32MB are executed in single transaction, that dramatically reduces overhead and lock contention.  Since the borrowing mechanism is not used for file rewrites, and it was never used by zvols, which seem to work fine, I don't think this change should create significant problems, partially because in addition to the borrowing mechanism there are also used pre-faults.

My tests with 4/8 threads writing several files same time on datasets with 32KB recordsize in 1MB requests show reduction of CPU usage by the user threads by 25-35%.  I would measure it in GB/s, but at that block size we are now limited by the lock contention of single write issue taskqueue, which is a separate problem we are going to investigate next.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: module/zfs/zfs_vnops.c [hunk 1]
```diff
@@ -544,29 +542,58 @@ zfs_write(znode_t *zp, zfs_uio_t *uio, int ioflag, cred_t *cr)
...
+		}
+
 		arc_buf_t *abuf = NULL;
-		if (n >= max_blksz && woff >= zp->z_size &&
-		    P2PHASE(woff, max_blksz) == 0 &&
-		    zp->z_blksz == max_blksz) {
+		ssize_t nbytes = n;
+		if (n >= blksz && woff >= zp->z_size &&
+		    P2PHASE(woff, blksz) == 0 &&
+		    (blksz >= SPA_OLD_MAXBLOCKSIZE || n < 4 * blksz)) {
```
Comment (bwatkinson) [inline module/zfs/zfs_vnops.c:569]: @amotin I am probably just not thinking through this correctly, but I am bit confused on this check. Is it possible for `n >= blksz` and `n < 4 * blksz` to ever be true at the same time?

Comment (amotin) [reply module/zfs/zfs_vnops.c:569]: @bwatkinson Yes. What is the problem?  For example, blksz == 16384 and n == 32748.

Comment (bwatkinson) [reply module/zfs/zfs_vnops.c:569]: Ah okay. Makes sense. Sorry for my confusion on this and thank you for clearing it up.

**Review by behlendorf — APPROVED** [review]

**Comment by ahrens** [comment]
We noticed a performance regression that seems to be caused by this change.  I think what's happening is that when we use bigger transactions, that applies to the ZIL transactions as well, and we may now call `zfs_log_write()` with a larger `tx_bytes`/`resid` than before.  That could cause us to use `WR_INDIRECT` whereas before we were using `WR_COPIED` or `WR_NEED_COPY`.  (e.g. with `recordsize=8k` and zfs_write() size of at least 32KB)

With WR_COPIED, the data was written to both the ZIL blocks, which are typically 128KB, and use `ZIO_PRIORITY_SYNC_WRITE`, and also to the data blocks as part of spa_sync(), which uses `ZIO_PRIORITY_ASYNC_WRITE`.  The spa_sync() writes can be aggregated, resulting in high average i/o size to disk (~90KB for us).

Now with WR_INDIRECT, there is less write inflation, but we use `ZIO_PRIORITY_SYNC_WRITE` for all the small blocks that constitute the bulk of the workload.  These writes aren't aggregated, resulting in many more i/o's to disk and worse overall performance (~3.5KB average write size to disk).

@amotin I wanted to see if this analysis makes sense to you, and if you have any thoughts on what behavior might be preferable.  We can try increasing `zfs_immediate_write_sz` to use WR_COPIED again on our benchmark, but that would also change the behavior of sync writes with recordsize=128k, which might not be desirable.  Another approach would be to try to do aggregation of `ZIO_PRIORITY_ASYNC_WRITE` i/os.

**Comment by amotin** [comment]
@ahrens I don't see us setting ZIO_FLAG_DONT_AGGREGATE for ZIO_PRIORITY_SYNC_WRITE, so in theory they should be aggregated, may be just less aggressively, since vdev_sync_write_max_active is higher than async.  But you may not have deep enough queue of them to reach that threshold, so it ends up up to device to handle multiple parallel small writes.

But otherwise, yes, switch from WR_COPIED to WR_INDIRECT is quite likely.  Though now the algorithm has more realistic input data so can be improved.  Actually zvol_log_write() unlike zfs_log_write() does not allow too small WR_INDIRECT blocks.  May be they could be just synchronized, unless there are better ideas.

**Comment by ahrens** [comment]
Thanks, that makes sense.  I think ZIO_PRIORITY_SYNC_WRITE being `vdev_queue_class_fifo()` contributes to the poor aggregation.  Maybe we will try changing that to see what effect it has (although being non-FIFO will probably hurt latency).

> Actually zvol_log_write() unlike zfs_log_write() does not allow too small WR_INDIRECT blocks. May be they could be just synchronized, unless there are better ideas.

Yes, we could change the meaning of `zfs_immediate_write_sz` to apply to the block size (`recordsize` property) rather than write syscall size.  This would be similar to how `zvol_immediate_write_sz` works.

**Comment by amotin** [comment]
> I think ZIO_PRIORITY_SYNC_WRITE being `vdev_queue_class_fifo()` contributes to the poor aggregation. Maybe we will try changing that to see what effect it has (although being non-FIFO will probably hurt latency).

It can make a difference only once you reach the requests threshold.  And even then the effect will be only if allocations are not sequential, that I somewhat doubt.

> Yes, we could change the meaning of `zfs_immediate_write_sz` to apply to the block size (`recordsize` property) rather than write syscall size. This would be similar to how `zvol_immediate_write_sz` works.

I was actually thinking that it could have sense to have two separate thresholds, one for block/record size and one for the total write size (effectively a number of indirects per transaction).  First would control maximum per-I/O overhead, while the second cover some of latency by writing and flushing multiple blocks same time.  But I haven't really benchmarked it, primarily testing pools with SLOG so far.

**Pull request merged**